crimson/seastore: refactor gc into concurrent task

athanatos opened the pull request.

athanatos commented:
@liu-chunmei This seems to have fixed several crashes exposed by crimson-store-nbd if you'd like to have a look.

tchaikov commented:
jenkins test make check

athanatos commented:
Missed those tests, will fix.

liu-chunmei commented on src/crimson/os/seastore/transaction_manager.cc:
```diff
@@ -274,69 +290,80 @@ TransactionManager::get_extent_if_live_ret TransactionManager::get_extent_if_liv
+      return lba_manager->get_mapping(
+	t,
+	laddr,
+	len).safe_then([=, &t](lba_pin_list_t pins) {
```
indent?

athanatos replied:
github is displaying the tabs incorrectly, it's fine.

liu-chunmei commented on src/crimson/os/seastore/transaction_manager.cc:
```diff
@@ -274,69 +290,80 @@ TransactionManager::get_extent_if_live_ret TransactionManager::get_extent_if_liv
+	type,
+	addr,
+	laddr,
+	len);
```
indent?

athanatos replied:
github is displaying the tabs incorrectly, it's fine.

athanatos commented on src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc:
```diff
@@ -522,7 +550,11 @@ BtreeLBAManager::update_refcount_ret BtreeLBAManager::update_refcount(
       return out;
     }).safe_then([](auto result) {
-      return ref_update_result_t{result.refcount, result.paddr};
+      return ref_update_result_t{
```
github is displaying the tabs incorrectly, it's fine.

athanatos commented on src/crimson/os/seastore/lba_manager/btree/lba_btree_node_impl.cc:
```diff
@@ -169,6 +169,11 @@ LBAInternalNode::mutate_internal_address_ret LBAInternalNode::mutate_internal_ad
     }
     auto iter = get_containing_child(laddr);
     if (iter->get_key() != laddr) {
+      logger().debug(
```
github is displaying the tabs incorrectly, it's fine.

athanatos commented on src/crimson/os/seastore/segment_cleaner.cc:
```diff
@@ -197,91 +199,92 @@ void SegmentCleaner::close_segment(segment_id_t segment)
+  return seastar::do_until(
+    [this] { return stopping; },
+    [this] {
+      return maybe_wait_should_run(
```
github is displaying the tabs incorrectly, it's fine.

athanatos commented on src/crimson/os/seastore/segment_cleaner.cc:
```diff
@@ -290,53 +293,67 @@ SegmentCleaner::do_gc_ret SegmentCleaner::do_gc(
-	    return ExtentCallbackInterface::release_segment_ertr::now();
-	  });
+	});
       });
```
github is displaying the tabs incorrectly, it's fine.

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -521,22 +539,100 @@ class SegmentCleaner : public JournalSegmentProvider {
+
+    gc_cycle_ret stop() {
+      if (!process_join)
+	return seastar::now();
```
github is displaying the tabs incorrectly, it's fine.

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -521,22 +539,100 @@ class SegmentCleaner : public JournalSegmentProvider {
+    void maybe_wake_on_space_used() {
+      cleaner.log_gc_state("GCProcess::maybe_wake_on_space_used");
+      if (cleaner.gc_should_run()) {
+	wake();
```
github is displaying the tabs incorrectly, it's fine.

athanatos commented on src/crimson/os/seastore/transaction_manager.cc:
```diff
@@ -158,7 +160,9 @@ TransactionManager::ref_ret TransactionManager::dec_ref(
 	offset);
-      return cache->retire_extent_if_cached(t, result.addr).safe_then([] {
+      return cache->retire_extent_if_cached(
+	t, result.addr, result.length
```
github is displaying the tabs incorrectly, it's fine.

athanatos commented:
@liu-chunmei Each of these files has

```
// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
// vim: ts=8 sw=2 smarttab
```

at the top setting offset and tab width as with most files in the ceph ptoject.  github isn't using the same tab width, so it frequently displays tabs in a way that throws off indents.

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -601,57 +713,111 @@ class SegmentCleaner : public JournalSegmentProvider {
+	journal_tail_target,
+	get_dirty_tail(),
+	get_dirty_tail_limit(),
+	gc_should_trim_journal()
```
Most already have comments at their declarations, I'll comment the others.

athanatos commented on src/crimson/os/seastore/segment_cleaner.cc:
```diff
@@ -290,53 +293,67 @@ SegmentCleaner::do_gc_ret SegmentCleaner::do_gc(
+			addr,
+			*ext);
+		    }
+		    return ecb->rewrite_extent(
```
Ok

athanatos commented:
Restricting the total size of journal entries to rewrite seems to have fixed the journal submission crash.  I'm debugging something else now, will push a fix when I figure it out.

liu-chunmei commented:
> Restricting the total size of journal entries to rewrite seems to have fixed the journal submission crash. I'm debugging something else now, will push a fix when I figure it out.

sure, I have tested the current PR, send you the debug log.

athanatos commented:
@liu-chunmei The hang I was debugging turned out to be because of a bad_alloc in segment_write -- I'd used --memory 64 instead of --memory 64G so it was using the default memory pool size.  It seems to be stable in my environment at the moment, so I'd like to go ahead and merge this series and take further bugs in a new PR.

athanatos commented:
I also added an explicit log line for that bad alloc and converted assert() to ceph_assert() in btree_range_pin -- that should make further problems easier to track down.

liu-chunmei commented:
> @liu-chunmei The hang I was debugging turned out to be because of a bad_alloc in segment_write -- I'd used --memory 64 instead of --memory 64G so it was using the default memory pool size. It seems to be stable in my environment at the moment, so I'd like to go ahead and merge this series and take further bugs in a new PR.

sure, let me test again to make clear the current status. We can merge it first and fix later.

liu-chunmei reviewed (commented):
Not every time segfault happen when set numjobs=1 and iodepth = xx.(sometimes success, and sometimes fail)  But when set numjobs =2, segfault always happen only once (iodepth =1) success.  no error log output when segfault.  we can fix it later since it can run and not block the further develop work.
